subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/OutgoingVariantsReportTask.java

```diff
@@ -0,0 +1,292 @@
+            tree.section("Attributes", () -> {
+                Integer max = attributes.keySet().stream().map(attr -> attr.getName().length()).max(Integer::compare).get();
+                attributes.keySet().stream().sorted(Comparator.comparing(Attribute::getName)).forEach(attr -> {
+                    tree.value(StringUtils.rightPad(attr.getName(), max), String.valueOf(attributes.getAttribute(attr)));
+                });
+            });
+            return true;
+        }
+        return false;
+    }
+
+    private void reportNoMatchingVariant(List<Configuration> configurations, StyledTextOutput output) {
+        if (variantSpec.isPresent()) {
+            output.println("There is no variant named '" + variantSpec.get() + "' defined on this project.");
+            configurations = getAllConsumableConfigurations().collect(Collectors.toList());
+        }
+        String projectName = getProject().getName();
+        if (configurations.isEmpty()) {
+            output.println("There are no outgoing variants on project " + projectName);
+        } else {
+            output.println("Here are the available outgoing variants: " + configurations.stream().map(Configuration::getName).collect(Collectors.joining(", ")));
+        }
+    }
+
+    private List<Configuration> filterConfigurations() {
+        Stream<Configuration> configurations = getAllConsumableConfigurations();
+        if (!showAll.get() && !variantSpec.isPresent()) {
+            configurations = configurations.filter(files -> !files.isCanBeResolved());
+        }
+        if (variantSpec.isPresent()) {
+            String variantName = variantSpec.get();
+            configurations = configurations.filter(cnf -> {
+                return cnf.getName().equals(variantName);
+            });
+        }
+        return configurations.collect(Collectors.toList());
+    }
+
+    private Stream<Configuration> getAllConsumableConfigurations() {
+        return getProject().getConfigurations()
+            .stream()
+            .filter(Configuration::isCanBeConsumed)
+            .sorted(Comparator.comparing(Configuration::getName));
+    }
+
+    private static class Legend {
+        private boolean hasPublications;
+        private boolean hasLegacyConfigurations;
+
+        void print(StyledTextOutput output) {
+            StyledTextOutput info = output.style(StyledTextOutput.Style.Info);
+            if (hasLegacyConfigurations || hasPublications) {
+                info.println();
+            }
+            if (hasLegacyConfigurations) {
+                info.println("(l) Legacy or deprecated configuration. Those are variants created for backwards compatibility which are both resolvable and consumable.");
+            }
+            if (hasPublications) {
+                info.text("(*) Secondary variants are variants created via the ")
+                .style(StyledTextOutput.Style.Identifier)
+                .text("Configuration#getOutgoing(): ConfigurationPublications")
```

Comment:
Maybe the type (`: ConfigurationPublications`) can be dropped here to make it shorter and more readable.

Reply:
This is technically the most important part: we work on a `ConfigurationPublication`, that you get from calling `#getOutgoing()`. That said I'm not sure it's important to make this shorter, this is really about debugging and plugin authors, so the more the better.